src/sinks/tcp.rs

```diff
@@ -106,30 +247,59 @@ impl TcpSink {
                             message = "disconnected.",
                             addr = &field::display(&self.addr)
                         );
-                        self.state = TcpSinkState::Disconnected;
+                        TcpSinkState::Disconnected
                     }
                 },
                 TcpSinkState::Connecting(ref mut connect_future) => match connect_future.poll() {
                     Ok(Async::Ready(socket)) => {
                         let addr = socket.peer_addr().unwrap_or(self.addr);
                         debug!(message = "connected", addr = &field::display(&addr));
-                        self.state =
-                            TcpSinkState::Connected(FramedWrite::new(socket, BytesCodec::new()));
                         self.backoff = Self::fresh_backoff();
+                        match self.tls {
+                            Some(ref tls) => match tls.make_connector() {
+                                Ok(connector) => TcpSinkState::TlsConnecting(
+                                    connector.connect(&self.hostname, socket),
+                                ),
+                                Err(err) => {
+                                    error!("{}", err);
+                                    TcpSinkState::Backoff(self.next_delay())
+                                }
+                            },
+                            None => TcpSinkState::Connected(Box::new(FramedWrite::new(
+                                socket,
+                                BytesCodec::new(),
+                            ))),
+                        }
                     }
                     Ok(Async::NotReady) => {
                         return Ok(Async::NotReady);
                     }
                     Err(err) => {
                         error!("Error connecting to {}: {}", self.addr, err);
-                        let delay = Delay::new(Instant::now() + self.backoff.next().unwrap());
-                        self.state = TcpSinkState::Backoff(delay);
+                        TcpSinkState::Backoff(self.next_delay())
                     }
                 },
+                TcpSinkState::TlsConnecting(ref mut connect_future) => {
+                    match connect_future.poll() {
+                        Ok(Async::Ready(socket)) => {
+                            debug!(message = "negotiated TLS");
+                            self.backoff = Self::fresh_backoff();
+                            TcpSinkState::Connected(Box::new(FramedWrite::new(
+                                socket,
+                                BytesCodec::new(),
+                            )))
+                        }
+                        Ok(Async::NotReady) => return Ok(Async::NotReady),
+                        Err(err) => {
+                            error!("Error negotiating TLS with {}: {}", self.addr, err);
```

Comment:
```suggestion
                            error!(message = "unable to negotiate TLS.", addr = %self.addr, error = %err);
```

Reply:
Fixed (though in my defence, I was just copying the pattern a few lines above this).